### DISCUSSION TITLE
Deleted Contracts - What happens to storage?

### DISCUSSION THREAD

**Discussion by TrevorJTClarke** [near/nearcore#3620] [Q&A] [answered]
I haven't fully tested here, so please forgive me. :)

I am able to deploy a contract with storage allocated, think guest book example. 
After a few txns sending new data to chain, I delete contract.

Is contract + all previous storage values removed or marked for removal?
Is contract name released to public for claiming? (guest-book.testnet gets deleted, now anyone can get it)

Hoping to get some thoughts around this, so I can solidify contract upgrade approaches.

**Comment by bowenwang1996** [comment, marked as answer]
If by "delete contract" you mean you deleted the account, then yes, everything in the contract storage is removed and the name can be reclaimed. For contract upgradability, there is some discussion here https://github.com/nearprotocol/NEPs/pull/123

**Reply by TrevorJTClarke** [reply]
Yes, that's what I meant. So basically root level account removal is susceptible to name sniping and should be avoided when possible.
Was thinking best approach is never deploy to: `*.near`
Basically always setup: `vX.base.near`
For example

**Reply by bowenwang1996** [reply]
I think it depends. If your contract operates permissionlessly, i.e, you don't have any access key and there is no way for the contract to add keys or delete itself, then you don't have to worry about it.
